site/app/libraries/Access.php

```diff
@@ -0,0 +1,248 @@
+        }
+        $checks = $this->permissions[$action];
+
+        //Some things may be available when there is no user
+        $user = $this->core->getUser();
+        if ($user === null) {
+            if (!($checks & self::ALLOW_LOGGED_OUT)) {
+                return false;
+            }
+            $group = User::GROUP_NONE;
+        } else {
+            $group = $user->getGroup();
+        }
+
+        //Check user group first
+        if ($group === User::GROUP_STUDENT && !($checks & self::ALLOW_STUDENT)) {
+            return false;
+        } else if ($group === User::GROUP_LIMITED_ACCESS_GRADER && !($checks & self::ALLOW_LIMITED_ACCESS_GRADER)) {
+            return false;
+        } else if ($group === User::GROUP_FULL_ACCESS_GRADER && !($checks & self::ALLOW_FULL_ACCESS_GRADER)) {
+            return false;
+        } else if ($group === User::GROUP_INSTRUCTOR && !($checks & self::ALLOW_INSTRUCTOR)) {
+            return false;
+        }
+
+        if ($checks & self::CHECK_CSRF) {
+            if ($this->core->checkCsrfToken()) {
+                return false;
+            }
+        }
+
+        if ($checks & self::CHECK_GRADEABLE_MIN_GROUP) {
+            /* @var Gradeable|null $gradeable */
+            $gradeable = $this->requireArg($args, "gradeable");
+            if ($gradeable === null) {
+                return false;
+            }
+            //Make sure they meet the minimum requirements
+            if (!$this->checkGroupPrivilege($group, $gradeable->getMinimumGradingGroup())) {
+                //You may be allowed to see this if you are trying to peer grade. Otherwise, you're not allowed
+                if (!($group === User::GROUP_STUDENT && $gradeable->getPeerGrading())) {
+                    return false;
+                }
+            }
+        }
+
+        if ($checks & self::CHECK_HAS_SUBMISSION) {
+            /* @var Gradeable|null $gradeable */
+            $gradeable = $this->requireArg($args, "gradeable");
+            if ($gradeable === null) {
+                return false;
+            }
+            if ($gradeable->getActiveVersion() <= 0) {
+                return false;
+            }
+        }
+
+        if ($group === User::GROUP_LIMITED_ACCESS_GRADER && ($checks & self::CHECK_GRADING_SECTION_GRADER)) {
+            /* @var Gradeable|null $gradeable */
+            $gradeable = $this->requireArg($args, "gradeable");
+            if ($gradeable === null) {
+                return false;
+            }
+            //Check their grading section
+            if (!$this->checkGradingSection($gradeable)) {
+                return false;
+            }
+        }
+        if ($group === User::GROUP_STUDENT && ($checks & self::CHECK_PEER_ASSIGNMENT_STUDENT)) {
+            /* @var Gradeable|null $gradeable */
+            $gradeable = $this->requireArg($args, "gradeable");
+            if ($gradeable === null) {
+                return false;
+            }
```

Comment:
This doesn't really feel like a great paradigm as if we wanted to add a new arg, we're just copying these four lines of boilerplate along and just continuously adding to this one function (which is already almost ~90 lines long). I guess you could do a check of `if ($checks & self:: CHECK_GRADEABLE_MIN_GROUP || $checks & self:: CHECK_HAS_SUBMISSION || ...)` and if so, enter a block that then checks, do we have a gradeable arg, if not, return false. You could do something with a bit flag, to also tell us to go into a "gradeable" specific check function.

Reply:
Yeah I was struggling with how to both require arguments for specific checks but then ignore them for others. A generic bit flag sounds like a pretty good way to go.